File: Retos/Reto #9 - HETEROGRAMA, ISOGRAMA Y PANGRAMA [Fácil]/javascript/cesar-ch.js
/*
 * Crea 3 funciones, cada una encargada de detectar si una cadena de
 * texto es un heterograma, un isograma o un pangrama.
 * - Debes buscar la definición de cada uno de estos términos.
 */

function isHeterograma(word) {
    let arrWords = []
    word.split('').map(e => {
        if (!arrWords.includes(e)) {
            arrWords.push(e)
        }
    })
    return arrWords.length === word.length
}

function isIsograma(word) {
    let objWords = {}
    word.split('').map(e => {
        if (objWords[e]) {
            objWords[e]++
        } else {
            objWords[e] = 1
        }
    })
    return Object.values(objWords).every(e => e % 2 === 0 || e === 1)
}

function isPangrama(word) {
    const alphabet = 'abcdefghijklmnñopqrstuvwxyz'
    let arrWords = []
    word.split('').map(e => {
        if (!arrWords.includes(e.toLowerCase()) && alphabet.split('').includes(e.toLowerCase())) {
            arrWords.push(e.toLowerCase())
        }
    })
    return arrWords.length === 27
}

console.log(isHeterograma('luteranismo'))
console.log(isIsograma('papelera'))
console.log(isPangrama('Benjamín pidió una bebida de kiwi y fresa. Noé, sin vergüenza, la más exquisita champaña del menú.'))

File: Retos/Reto #9 - HETEROGRAMA, ISOGRAMA Y PANGRAMA [Fácil]/python/cesar-ch.py
"""
 * Crea 3 funciones, cada una encargada de detectar si una cadena de
 * texto es un heterograma, un isograma o un pangrama.
 * - Debes buscar la definición de cada uno de estos términos.
"""


def is_heterograma(word):
    arr_words = []
    for e in word:
        if e not in arr_words:
            arr_words.append(e)
    return len(arr_words) == len(word)


def is_isograma(word):
    obj_words = {}
    for e in word:
        if e in obj_words:
            obj_words[e] += 1
        else:
            obj_words[e] = 1
    return all(e % 2 == 0 or e == 1 for e in obj_words.values())


def is_pangrama(word):
    alphabet = 'abcdefghijklmnñopqrstuvwxyz'
    arr_words = []
    for e in word:
        if e.lower() not in arr_words and e.lower() in alphabet:
            arr_words.append(e.lower())
    return len(arr_words) == 27


print(is_heterograma('luteranismo'))
print(is_isograma('papelera'))
print(is_pangrama('Benjamín pidió una bebida de kiwi y fresa. Noé, sin vergüenza, la más exquisita champaña del menú.'))
